From 003be038764aa9e3cfc819a37c679e9412da56da Mon Sep 17 00:00:00 2001
From: KayYCHuang-wiwynn <kay_yc_huang@wiwynn.com>
Date: Tue, 24 Jun 2025 10:12:30 +0000
Subject: [PATCH 19/19] Add event log type for PROCHOT

[Task Description]
- Related to YV4T1M-1031
- Develop a mechanism to trigger PROCHOT when thermal sensors exceed their upper critical threshold
- Implement event logging for PROCHOT events initiated by thermal sensor readings

[Motivation]
- Enhance system thermal protection by proactively triggering PROCHOT during thermal overload scenarios

[Design]
- Add a new event type "PROCHOT is triggered due to SD sensor reading exceed UCR"

Signed-off-by: KayYCHuang-wiwynn <kay.yc.huang.wiwynn@gmail.com>
---
 .../oem_meta_file_io_type_event_log.cpp       | 32 +++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp b/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp
index d7e3fa6..d20ca08 100644
--- a/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp
+++ b/oem/meta/libpldmresponder/oem_meta_file_io_type_event_log.cpp
@@ -1,7 +1,9 @@
 #include "oem_meta_file_io_type_event_log.hpp"
 
+#include <phosphor-logging/commit.hpp>
 #include <phosphor-logging/lg2.hpp>
 #include <sdbusplus/bus.hpp>
+#include <xyz/openbmc_project/State/Thermal/event.hpp>
 
 #include <utility>
 
@@ -103,6 +105,7 @@ int EventLogHandler::write(const message& eventData)
         " CXL2 HB",
         " Post-Started",
         " Post-Ended",
+        " PROCHOT is triggered due to SD sensor reading exceed UCR"
     });
 
     static constexpr auto pmic_event_type = std::to_array({
@@ -295,6 +298,35 @@ int EventLogHandler::write(const message& eventData)
             error("Failed to create info-level D-Bus log: {ERR}", "ERR", e);
         }
     }
+    else if (event_name ==
+             " PROCHOT is triggered due to SD sensor reading exceed UCR")
+    {
+        std::string device = "/xyz/openbmc_project/State/Thermal/host" +
+                             std::to_string(slot) + "/cpu0";
+        auto oemData = "0x" +
+                       std::format("{:02x}{:02x}", eventData[3], eventData[4]);
+        std::string failureData =
+            "SD sensor reading exceed UCR, data: " + oemData;
+
+        if (oemData == "0x0000")
+        {
+            using DeviceNormalTemp = sdbusplus::event::xyz::openbmc_project::
+                state::Thermal::DeviceOperatingNormalTemperature;
+            lg2::commit(DeviceNormalTemp(
+                DeviceNormalTemp::metadata_t<"DEVICE">{"DEVICE"},
+                sdbusplus::message::object_path(device)));
+        }
+        else
+        {
+            using DeviceOverTemp = sdbusplus::error::xyz::openbmc_project::
+                state::Thermal::DeviceOverOperatingTemperature;
+            lg2::commit(DeviceOverTemp(
+                DeviceOverTemp::metadata_t<"DEVICE">{"DEVICE"},
+                sdbusplus::message::object_path(device),
+                DeviceOverTemp::metadata_t<"FAILURE_DATA">{"FAILURE_DATA"},
+                failureData));
+        }
+    }
     else
     {
         lg2::error("{ERROR}", "ERROR", errorLog); // Create log in journal
-- 
2.47.1

